src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1388 @@
+
+    void Clear() override
+    {
+        keyset_id     = 0;
+        policy        = Keyset::SecurityPolicy::kStandard;
+        num_keys_used = 0;
+        next          = 0;
+        memset(epoch_keys, 0x00, sizeof(epoch_keys));
+    }
+
+    CHIP_ERROR Serialize(TLV::TLVWriter & writer) const override
+    {
+        // keyset_id
+        ReturnErrorOnFailure(writer.Put(kTagKeysetId, static_cast<uint16_t>(keyset_id)));
+        // policy
+        ReturnErrorOnFailure(writer.Put(kTagPolicy, static_cast<uint16_t>(policy)));
+        // num_keys_used
+        ReturnErrorOnFailure(writer.Put(kTagNumKeys, static_cast<uint16_t>(num_keys_used)));
+        // epoch_keys
+        {
+            TLV::TLVType array, item;
+            writer.StartContainer(kTagEpochKeys, TLV::kTLVType_Array, array);
+            for (auto & epoch : this->epoch_keys)
+            {
+                writer.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, item);
+                ReturnErrorOnFailure(writer.Put(TLV::ContextTag(kTagStartTime), static_cast<uint64_t>(epoch.start_time)));
+                ReturnErrorOnFailure(
+                    writer.PutBytes(TLV::ContextTag(kTagKey), epoch.key, GroupDataProvider::EpochKey::kLengthBytes));
+                writer.EndContainer(item);
+            }
+            writer.EndContainer(array);
+        }
+        // next keyset
+        ReturnErrorOnFailure(writer.Put(kTagNext, static_cast<uint16_t>(next)));
+
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR Deserialize(TLV::TLVReader & reader) override
+    {
+        // keyset_id
+        ReturnErrorOnFailure(reader.Next(kTagKeysetId));
+        ReturnErrorOnFailure(reader.Get(keyset_id));
+        // policy
+        ReturnErrorOnFailure(reader.Next(kTagPolicy));
+        ReturnErrorOnFailure(reader.Get(policy));
+        // num_keys_used
+        ReturnErrorOnFailure(reader.Next(kTagNumKeys));
+        ReturnErrorOnFailure(reader.Get(num_keys_used));
+        {
+            // epoch_keys
+            ReturnErrorOnFailure(reader.Next(kTagEpochKeys));
+            VerifyOrReturnError(TLV::kTLVType_Array == reader.GetType(), CHIP_ERROR_INTERNAL);
+
+            TLV::TLVType array, item;
+            reader.EnterContainer(array);
+            for (auto & epoch : this->epoch_keys)
+            {
+                ReturnErrorOnFailure(reader.Next(TLV::AnonymousTag));
+                VerifyOrReturnError(TLV::kTLVType_Structure == reader.GetType(), CHIP_ERROR_INTERNAL);
+
+                reader.EnterContainer(item);
+                // start_time
+                ReturnErrorOnFailure(reader.Next(TLV::ContextTag(kTagStartTime)));
+                ReturnErrorOnFailure(reader.Get(epoch.start_time));
+                // key
+                ReturnErrorOnFailure(reader.Next(TLV::ContextTag(kTagKey)));
+                ReturnErrorOnFailure(reader.GetBytes(epoch.key, GroupDataProvider::EpochKey::kLengthBytes));
```

Comment:
This can read less than expected. A better way is to `CHIP_ERROR Get(ByteSpan & v);` into a ByteSpan, and then validate the span's `.size()` matches how much you expect to read out of a key

Reply:
Fixed as suggested.